### DISCUSSION TITLE
Behavior for out-of-bounds slice indices should be defined

### DISCUSSION THREAD

**Discussion by gibson042** [jmespath-community/jmespath.spec#122] [General]
See [discussion](https://github.com/jmespath-community/jmespath.spec/pull/102#discussion_r1008913623) in #102.

Behavior is currently undefined for out-of-bounds slice indices, e.g. ``` `[]`[10:] ``` and `'foo'[-100:100]`. The former at least should probably align with Python and be valid but clamp indices, whereas the latter could either align with the former **or** raise exceptions (ideally making sense with whatever the decision is for out-of-bounds string index expressions such as `'foo'[-1]` and `'foo'[100]`, which are also currently undefined).

**Comment by springcomp** [comment]
I agree that the spec lacks precision in this regards but I think the spirit is to align with Python as per:

_Slices in JMESPath have the same semantics as python slices. If you're familiar with python slices, you're familiar with JMESPath slices._

In Python slice indices are not clamped so that explains most of the behaviour that we see.
I think it is natural for string slices to follow this pattern, as is already the case for Python string slices.

**Reply by gibson042** [reply]
Great; I think we agree, although I describe the behavior as clamping because any index that is out of bounds because it is too low or too high is instead treated as if it were respectively the minimum or maximum valid value (as opposed to e.g. rejecting the range or synthesizing substitute/default collection items).

**Reply by gibson042** [reply]
And to provide some concrete examples:
* ``` `[0, 1, 2]`[-100:100]``` results in `[0, 1, 2]`
* `'foo'[-100:100]` results in `"foo"`
* ``` `[0, 1, 2]`[-100:1]``` results in `[0]`
* `'foo'[-100:]` results in `"foo"`
* `'foo'[:100]` results in `"foo"`
* `'foo'[-100:1]` results in `"f"`
